packages/grafana-ui/src/components/Monaco/CodeEditor.tsx

```diff
@@ -54,9 +55,21 @@ class UnthemedCodeEditor extends React.PureComponent<Props> {
       if (getSuggestions) {
         this.completionCancel = registerSuggestions(this.monaco, language, getSuggestions);
       }
+
+      this.loadCustomLanguage();
     }
   }
 
+  loadCustomLanguage = () => {
+    const { language } = this.props;
+
+    if (language === 'kusto') {
+      return loadKusto();
+    }
+
+    return Promise.resolve();
+  };
+
```

Comment:
I'm not sure languages should live in the grafana/ui.

IMO, an alternative solution to that would be for grafana/ui to expose an interface (registry to be specific, similar to editors registry https://github.com/grafana/grafana/blob/0cc620aea7417c3d594aab29d4d86c115d5fad36/packages/grafana-data/src/field/standardFieldConfigEditorRegistry.ts#L28, that would allow dynamic languages registration.

This way core Grafana could come with languages built in (think kusto for example) and plugins could hook into this registry by providing their own languages.

custom language loader method would then turn into

```
loadCustomLanguage = () => {
    const { language } = this.props;

   const languageItem = monacoLanguageRegistry.get('language');
   if(languageItem) {
     return languageItem.loadDefinition() 
   }

  return Promise.resolve()
  };
```

Reply:
I think I agree in concept, but I'm not sure if I follow in the specifics of implementation.

Where should the kusto.ts loader live in the repo? Where would grafana core register the languages with grafana/ui?

Personally I think that language support for the code editor is a UI concern, not a core grafana concern. And consumers of <CodeEditor /> can provide their own language support just by using the `onEditorDidMount` hook (that's, after all, what this is doing here). But I don't feel too strongly on this.

Reply:
> Personally I think that language support for the code editor is a UI concern

the UI component should make sure loading custom components is possible.  If we need a new grafana/ui release to add a new language, we will be sad.

For this exact case of putting the library into core plugins the difference is kinda academic.  the kusto files should live in `public/lib/monaco/...` anyway, but we I agree with @dprokop that the specific "kusto" language support should be possible to initalize outside of grafana/ui.

Reply:
Monaco custom language support is just a matter of registering your language before/after the editor is mounted, and the <CodeEditor /> component already provides the lifecycle hooks to allow consumers to do this.

If you follow the code, there's no reason why `loadKusto()` could not just be called in the userland `editorDidMount` callback. The only reason I put it within <CodeEditor /> is because I want to reuse it in the ADX plugin.

Happy to take concrete actionable feedback for how to improve this, if we feel it needs it.

Reply:
@joshhunt as I and @ryantxu proposed:
1. Move kusto lang to grafana (public/lib/monaco...)
2. Create a registry that allows dynamic languages definition (`monacoLanguageRegistry`). I think it should live in grafana/data
3. In CodeEditor use the registry to retrieve and load the language rather than load the kusto lang specifically in the CodeEditor
4. The "default" languages need to be added to a registry - you can do it similarly to how standard editors are defined (https://github.com/grafana/grafana/blob/master/public/app/app.ts#L95)

I hope this helps

Reply:
Cool, thanks!

I've made these changes, would love thoughts on it.

I wasn't exactly sure about the point 1 - the monaco-kusto language files are already being copied into `public/lib/monaco/min/vs/language/kusto/`, but separately there is our code for how we load that. I've moved that into `public/lib/monaco-languages` along with the default registry values for it (keep in mind entire `public/lib/monaco` folder is .gitignored and copied out of node_modules by webpack)